Fatras/include/ActsFatras/Kernel/detail/Interactor.hpp

```diff
@@ -96,14 +100,26 @@ struct Interactor {
     // decay check. needs to happen at every step, not just on surfaces.
     // TODO limit the stepsize when close to the lifetime limit to avoid
     //   overstepping and decaying the particle systematically too late
-    if (result.properTimeLimit < result.particle.properTime()) {
+    if ((result.properTimeLimit - result.particle.properTime() < result.properTimeLimit * properTimeTolerance) &&
+        result.isAlive) {
       auto descendants = decay.run(generator, result.particle);
       for (auto &&descendant : descendants) {
         result.generatedParticles.emplace_back(std::move(descendant));
       }
       result.isAlive = false;
       return;
     }
+    
+    // Regulate the step size 
+    if(std::isfinite(result.properTimeLimit)) {
```

Comment:
This check is probably unnecessary. If the value is inf, than you already get an infinite step size from the calculations below.

Reply:
The thing is that you don't need to calculate the code below if your particle is stable

Reply:
Branching is quite expensive on modern CPUs. For small computation, it can be faster to execute them unconditionally. It is not so important here, so let's leave it as-is.

Reply:
FWIW, this kind of branch where you're always taking the same direction unless an exceptional event happens is not that bad at the CPU level. It should usually be well predicted, in which case it takes just 1 cycle (and a bit of branch predictor state).

Unless my profiler tells me otherwise, I only start worrying about branches when:

1. The branch condition follows a complex pattern that the CPU cannot easily predict.
2. The branch is guarding a very short and hot computation, in which case the extra cycle is still one cycle too much.
3. The branch is hurting the compiler's loop optimizations.